helix-core/src/text_annotations.rs

```diff
@@ -0,0 +1,183 @@
+use std::cell::Cell;
+use std::convert::identity;
+use std::ops::Range;
+
+use crate::graphemes::GraphemeStr;
+use crate::syntax::Highlight;
+
+#[derive(Debug)]
+pub struct InlineAnnotation {
+    pub text: Box<str>,
+    pub char_idx: usize,
+}
+
+#[derive(Debug)]
+pub struct Overlay<'t> {
+    pub char_idx: usize,
+    pub grapheme: GraphemeStr<'t>,
+}
+
+#[derive(Debug)]
+pub struct LineAnnotation {
+    pub anchor_char_idx: usize,
+    pub height: usize,
+}
+
+#[derive(Debug)]
+struct Layer<'a, A, M> {
+    annotations: &'a [A],
+    current_index: Cell<usize>,
+    metadata: M,
+}
+
+impl<'a, T, M: Clone> Clone for Layer<'a, T, M> {
+    fn clone(&self) -> Self {
+        Layer {
+            annotations: self.annotations,
+            current_index: self.current_index.clone(),
+            metadata: self.metadata.clone(),
+        }
+    }
+}
+
+impl<'a, T, M> Layer<'a, T, M> {
+    pub fn reset_pos(&self, char_idx: usize, get_char_idx: impl Fn(&T) -> usize) {
+        let new_index = self
+            .annotations
+            .binary_search_by_key(&char_idx, get_char_idx)
+            .unwrap_or_else(identity);
+
+        self.current_index.set(new_index);
+    }
+
+    pub fn consume(&self, char_idx: usize, get_char_idx: impl Fn(&T) -> usize) -> Option<&'a T> {
+        let annot = self.annotations.get(self.current_index.get())?;
+        debug_assert!(get_char_idx(annot) >= char_idx);
```

Comment:
As an API consumer all you need to care about is `add_overlay_layer` which accepts a list of grapheme `Overlay`s (not that it does not accept a char range you need to produce an individual overlay for each character/grapheme you are highlighting individually). Each `Overlay` contains a `char_idx` and the new `grapheme` to display instead of the old one. It is however important that the overlay layer you provide is **sorted by char_idx**.

The `TextAnnotations` and `DocumentCursor` will take care of the rest by automatically calling `TextAnnotations::reset_pos` when starting an iteration which will reset the internal position of `TextAnnotations` to the correct point using a binary search. This is an implementation detail tough

Reply:
I added a lot more documentation to the `text_overlay` module which should hopefully provide some clarity.
You mostly just want to store a `jump_lables: Vec<Overlay>`  either in the `View` or the `Document` which gets populated/cleaned up by the command and add these `Overlay`s as a layer in in the `text_additions` function.